docs/kns_apps_tutorial.md

```diff
@@ -0,0 +1,347 @@
+* `global_retry_sleep`: how much seconds should we wait between each retry.
+
+#### 2.3.2. Templates
+
+```json
+                "templates": [
+                    {
+                        "name": "google-sample",
+                        "type": "kubectl",
+                        "path": "google-sample.yml"
+                    }
+                ]
+```
+
+The `templates` list allow to execute instructions in an ordered manner, and to apply templates to deploy the app into the nested scheduler.
+Since the init pod runs with cluster-admin rights (inside the nested K3S cluster), templates can create new namespaces, cluster roles, etc.
+
+Note that before executing templates, the KNS will pull and extract the app image. All files inside the `/templates` folder of the image will be extracted and execution will take place **inside** this extracted folder.
+
+For example, in our current example, hello world, the dockerfile contains:
+
+```dockerfile
+RUN mkdir /templates
+COPY google-sample.yml /templates/google-sample.yml
+```
+
+For each step, a `name` parameter must be set. This is only for convenience, and ease logs reading.
+
+It is also possible to fine define retry and sleep between retries:
+
+```json
+                "templates": [
+                    {
+                        "name": "google-sample",
+                        "type": "kubectl",
+                        "path": "google-sample.yml",
+                        "retry": 10,
+                        "retry_sleep": 30
+                    }
+                ]
+```
+
+For this item, `retry` will precedence `global_retry` if set, and `retry_sleep` will precedence `global_retry_sleep`.
+
+There are then 3 types of actions available:
+
+##### 2.3.2.1. kubectl type
+
+Kubectl will execute the following command:
+
+```
+/usr/bin/kubectl apply -f path/to/template.yml
+```
+
+If `type` is set to `kubectl`, then a `path` must be provided, relative to the root of `/templates` folder of the app image.
+In our example, since our `google-sample.yml` file is at `/templates/google-sample.yml` into our app image, `path` will simply be `google-sample.yml`.
+
+Note that the kubectl type also allows to use an URL as a template, to fetch a remote template (with all risks that it implies!).
+To do so, specify a special key `format` to `url`, and provide the URL inside the `path` key. For example, to deploy ArgoCD:
+
+```json
+                "templates": [
+                    {
+                        "name": "allinone",
+                        "type": "kubectl",
+                        "path": "https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml",
+                        "format": "url"
+                    }
+                ]
+```
+
+Again, YOU MUST trust the owner of the remote template to use this feature.
+
+##### 2.3.2.1. kustomize type
+
+Kustomize will execute the following command:
+
+```
+/usr/bin/kubectl apply -k path/to/folder/
```

Comment:
Is this correct? Should not use the example of the command kustomize instead of kubectl?